ykrt/src/mt.rs

```diff
@@ -305,20 +363,58 @@ mod tests {
             t.join().unwrap();
         }
         {
-            mtt.control_point(&l_arc);
+            mtt.control_point(Some(&l_arc), dummy_step, ());
             assert_eq!(l_arc.pack.load(Ordering::Relaxed), PHASE_TRACING);
-            mtt.control_point(&l_arc);
+            mtt.control_point(Some(&l_arc), dummy_step, ());
             assert_eq!(l_arc.pack.load(Ordering::Relaxed), PHASE_COMPILED);
         }
     }
 
+    #[test]
+    fn dumb_interpreter() {
+        let mut mtt = MTBuilder::new().hot_threshold(2).init();
+
+        // The program is silly. Do nothing twice, then start again.
+        enum ByteCode {
+            Nop,
+            Restart,
+        }
+        let prog = vec![ByteCode::Nop, ByteCode::Nop, ByteCode::Restart];
+
+        // Suppose the bytecode compiler for this imaginary language knows that the first bytecode
+        // is the only place a loop can start.
+        let locs = vec![Some(Location::new()), None, None];
```

Comment:
Does this mean the locations vector is always the size of the input program bytecode?

Reply:
It doesn't have to be. The user can store them however they want. It could have been a hashmap, or any other sparse data structure.

(we probably don't want a hashmap, it was just an example)

Reply:
@ltratt What kind of data structure would you envisage for a production VM actually?

You'd need some kind of sparse mapping, but you said you want to avoid hashing. What would you use?

Reply:
Let's remember that there's no need to have a "locations vector": it's just one way (the simplest, probably; the least efficient, almost certainly) to encode a notion of locations. Ultimately the VM just has to give the meta-tracer a `Location` for every loop (or loop-like thing like a function).

A high-performance VM would probably do things like have a "spare" word after each loop header instruction to store `Location`s in.

Reply:
> A high-performance VM would probably do things like have a "spare" word after each loop header instruction to store Locations in.

Clever.